### PR TITLE
Adding MLIR cpp backend integration implementation

### PR DESCRIPTION
**Pull request by jerryyin**
This PR presents the cpp implementation for roadmap at https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/37. Implementation details are as follows:

* Use Jenkinsfile to pull in MLIR as a source component
  * Added ninja build as dependency to build MLIR
  * Added cmake logic to pull in libMLIRMIOpen
* Implemented three solvers one each for a independent direction
* Defined the interface needed for MLIR cpp backend
* Implemented the infrastrcture need to build MLIR kernels
* Added new test stage that specifically test mlir solvers
  * Added new test cases to test each solver

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -377,6 +377,46 @@ pipeline {
...
                 }
 
+                stage('Hip Implicitgemm layouts Release') {
+                    agent{ label rocmnode("vega") }
+                    environment{
+                        cmd = '''
+                            # check out MLIR
+                            dir=$(pwd)
+                            cd
+                            git clone -b miopen-dialect https://github.com/ROCmSoftwarePlatform/llvm-project-mlir.git
```
Comment (pfultz2) [inline Jenkinsfile:387]: This should be built and installed in the dockerfile. Also, this should use a commit hash and not a branch name.

File: Jenkinsfile [hunk 2]
```diff
@@ -377,6 +377,46 @@ pipeline {
...
+                            # config MLIR on ROCm, with MIOpen dialect
+                            cmake -G Ninja ../llvm \
+                              -DLLVM_ENABLE_PROJECTS="mlir" \
+                              -DCMAKE_BUILD_TYPE=Release \
+                              -DBUILD_SHARED_LIBS=OFF \
+                              -DLLVM_BUILD_LLVM_DYLIB=OFF \
+                              -DLLVM_ENABLE_TERMINFO=OFF
+
+                            # build libMLIRMIOpen
+                            cmake --build . --target libMLIRMIOpen
```
Comment (pfultz2) [inline Jenkinsfile:401]: Also this should be installed and then the source and build directory should be deleted.

File: hip-clang.docker [hunk 3]
```diff
@@ -28,6 +28,7 @@ RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --allow-
     libpthread-stubs0-dev \
     llvm-amdgpu \
     miopengemm \
+    ninja-build \
```
Comment (pfultz2) [inline hip-clang.docker:31]: Why is this necessary?

File: Jenkinsfile (hunk 1 above)
Comment (jerryyin) [reply Jenkinsfile:387]: Will building/installing inside the dockerfile cause it drag the build longer? I thought of the two alternatives a while ago and make it started in a jenkins file just so that non igemm stages do not suffer a performance penalty.

Besides, this review will not have everything in the final place at one go. Once the igemm got completely integrated, so that I can test not only NCHW layout, but also other layouts, I will move that to the dockerfile. Would this plan be okay?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -631,6 +634,16 @@ if(NOT WIN32 AND NOT APPLE)
     endif()
 endif()
 
+############################################################
+# MIOpen depends on libMLIRMIOpen for implicit gemm
+if(LIBMLIRMIOPEN)
+  MESSAGE(STATUS "LibMLIRMIOpen: " ${LIBMLIRMIOPEN})
+  set(LIBMLIRMIOPEN "${LIBMLIRMIOPEN}")
```
Comment (pfultz2) [inline src/CMakeLists.txt:641]: What is this for? This variable is never set. Shouldn't you call `find_library` or `find_package`(preferred) to find and link mlir?

File: Jenkinsfile (hunk 2 above)
Comment (jerryyin) [reply Jenkinsfile:401]: This goes away together with the docker. A few lines above, it create and build MLIR in the home directory, not the mounted jenkins workspace directory.

File: hip-clang.docker (hunk 3 above)
Comment (jerryyin) [reply hip-clang.docker:31]: `ninja-build` is the default (and faster) way to build LLVM project. If you are against pulling in this item, I can explore to try using cmake only.

File: Jenkinsfile (hunk 1 above)
Comment (pfultz2) [reply Jenkinsfile:387]: > Will building/installing inside the dockerfile cause it drag the build longer?

No, in fact it should be faster because the docker is cached on the nodes. Also, I assume its only temporary until this is upstreamed and provided by the llvmadgpu package.

> Besides, this review will not have everything in the final place at one go.

Yes, but this should be available for developers to run locally since this is being ran on jenkins.

File: src/CMakeLists.txt (hunk 4 above)
Comment (pfultz2) [reply src/CMakeLists.txt:641]: Nevermind, I see the `find_library` call.

Comment (pfultz2) [reply src/CMakeLists.txt:641]: This `set` still seems odd though.

File: hip-clang.docker (hunk 3 above)
Comment (pfultz2) [reply hip-clang.docker:31]: But I believe rocm just uses make to build the llvm package.

File: src/CMakeLists.txt (hunk 4 above)
Comment (jerryyin) [reply src/CMakeLists.txt:641]: I'm fairly new on cmake let me know if there's a better way to do it.

I used to have the `find_library` called in this place, but not in root directory's `CMakeLists.txt` as I do now. The problem is that I also reuse the `LIBMLIRMIOpen` as the `#define` variable. What that means is that `find_library()` call has to go before the 

```configure_file("{SOURCE_DIR}/include/config.h.in" "{BINARY_DIR}/include/config.h"). ```

Unfortunately that `configure_file` line is on the top of the `src/CMakeLists.txt`. That's why I ended up having the `find_library()` call in the root directory's CMakeLists.txt, and set it at a later stage, in here.

File: Jenkinsfile (hunk 2 above)
Comment (pfultz2) [reply Jenkinsfile:401]: It should be installed into /usr/local. Ideally, we should do `make package` and install this in place of the llvm package provided by rocm. It seems mismatching across different llvm versions could cause problems. Although, updating the llvm version will need adjustments or fixing with clang-tidy.

File: src/CMakeLists.txt (hunk 4 above)
Comment (pfultz2) [reply src/CMakeLists.txt:641]: You should use two different variable names. One for `find_library` and another for `configure_file`.

File: Jenkinsfile (hunk 1 above)
Comment (jerryyin) [reply Jenkinsfile:387]: > No, in fact it should be faster because the docker is cached on the nodes. 
> Yes, but this should be available for developers to run locally since this is being ran on jenkins.

Sure, I can move this logic into dockerfile

> Also, I assume its only temporary until this is upstreamed and provided by the llvmadgpu package.

This will probably happen in a longer period. We upstream zero percent of the `miopen-dialect` inside of MLIR and this static library is a fat library that includes all the necessary llvm dependencies together. Desirably when we finished upstreaming, this library can ship together with rest of llvm libraries.

File: include/miopen/config.h.in [hunk 5]
```diff
@@ -75,6 +75,7 @@
 #cmakedefine EXTRACTKERNEL_BIN "@EXTRACTKERNEL_BIN@"
 #cmakedefine MIOPEN_OFFLOADBUNDLER_BIN "@MIOPEN_OFFLOADBUNDLER_BIN@"
 #cmakedefine MIOPEN_CACHE_DIR "@MIOPEN_CACHE_DIR@"
+#cmakedefine LIBMLIRMIOPEN "@LIBMLIRMIOPEN@"
```
Comment (pfultz2) [inline include/miopen/config.h.in:78]: This should be prefixed with `MIOPEN_`

File: Jenkinsfile (hunk 2 above)
Comment (jerryyin) [reply Jenkinsfile:401]: This cmake build will result in calling of a `add_custom_command` on the post build stage that put the built library under `/opt/rocm/lib`. 

> we should do make package and install this in place of the llvm package provided by rocm

So yes, as you mentioned, this is a different llvm version than the one provided by rocm. Since everything is zipped in this one target fat static library, I think it is better to make it separate from the rocm provided llvm.

File: hip-clang.docker (hunk 3 above)
Comment (jerryyin) [reply hip-clang.docker:31]: I didn't know about rocm way to build llvm package. If I understand it right, there is an option to use cmake -G with different generators. ninja are the one we typically use because it is faster. I can explore the -G Unix Makefiles option.

File: src/CMakeLists.txt (hunk 4 above)
Comment (jerryyin) [reply src/CMakeLists.txt:641]: Will give that a shot, thanks!

File: include/miopen/config.h.in (hunk 5 above)
Comment (jerryyin) [reply include/miopen/config.h.in:78]: Will create a new variable, and prefix it with MIOPEN_ in here.

**Comment by jerryyin** [comment]
Closing the PR now, because some Jenkins changes result in left over root permission workspace. Will reopen in a new branch while addressing the permission issues in this branch.

**Comment by jerryyin** [comment]
Reopened in #676
